design/design-doc-packages-v2.md

```diff
@@ -0,0 +1,1067 @@
+
+### Package Controller
+
+The `Package` controller is responsible for watching the create / update /
+delete a `Package`, unpacking it, and creating the corresponding
+`PackageRevision`. It is knowledgeable of and modifies instances of both the
+`Package` and `PackageRevision` types. For Phase 1, the `Package` controller
+will operate with the following behavior.
+
+**`Package` Created with `spec.revisionActivationPolicy: Automatic`**
+
+1. Observe `spec.package` and create `Job` to unpack the package.
+2. When `Job` is complete, create `PackageRevision` from `Job` output with
+   `spec.desiredState: Active`.
+3. Set `status.currentRevision` to full image name used for `PackageRevision`
+   (this can be obtained from the `Pod` in the install `Job`)
+4. Report status of `PackageRevision` in the the status of the `Package`.
+
+**`Package` Created with `spec.revisionActivationPolicy: Manual`**
+
+1. Observe `spec.package` and create `Job` to unpack the package.
+2. When `Job` is complete, create `PackageRevision` from `Job` output with
+   `spec.desiredState: Inactive`.
+3. Set `status.currentRevision` to full image name used for `PackageRevision`
+   (this can be obtained from the `Pod` in the install `Job`)
+4. Report status of `PackageRevision` in the the status of the `Package`.
+
+User is responsible for manually setting the `PackageRevision` to `Active`.
+
+**Existing `Package` with `spec.revisionActivationPolicy: Automatic` Modified**
+
+1. Observe difference in `spec.package` and `status.currentRevision`. If an
+   install `Job` had previously failed to pull a package image and only the
+   `spec.imagePullSecrets` changed, the `Job` would be retried with the new
+   `spec.imagePullSecrets`.
+2. Create install `Job` to unpack new package version.
+3. When `Job` is complete, mark `Active` `PackageRevision` as
+   `spec.desiredState: Inactive`, and create new `PackageRevision` as
+   `spec.desiredState: Active`.
+4. Set `status.currentRevision` to full image name used for new
+   `PackageRevision` (this can be obtained from the `Pod` in the install `Job`).
+5. Clean up any install `Job` for any `PackageRevision` that has successfully
+   transitioned to `Inactive` state.
+
+**Existing `Package` with `spec.revisionActivationPolicy: Manual` Modified**
+
+1. Observe difference in `spec.package` and `status.currentRevision`. If an
+   install `Job` had previously failed to pull a package image and only the
+   `spec.imagePullSecrets` changed, the `Job` would be retried with the new
+   `spec.imagePullSecrets`.
+2. Create install `Job` to unpack new package version.
+3. When `Job` is complete, create new `PackageRevision` with `spec.desiredState:
+   Inactive`.
+4. Wait for user to set new `PackageRevision` to `spec.desiredState.Active`.
+5. Mark existing `PackageRevision` as `Inactive` and set
+   `status.currentRevision` to full image name used for new `PackageRevision`
+   (this can be obtained from the `Pod` in the install `Job`).
```

Comment:
> Will the package manager effectively set any active revision to inactive when it notices another revision transitioning to desiredState: Active?

It depends on the activation strategy. In the manual case, yes. In the automatic case, the package manager is responsible for determining which revision is supposed to be active based on the revision, so it could mark a revision as inactive before it creates the new revision and marks it as active.